src/java/org/apache/cassandra/service/HeartbeatService.java

```diff
@@ -0,0 +1,137 @@
+
+import static org.apache.cassandra.concurrent.ExecutorFactory.Global.executorFactory;
+import static org.apache.cassandra.service.GcGraceSecondsOnStartupCheck.getHeartbeatFile;
+import static org.apache.cassandra.service.StartupChecks.StartupCheckType.gc_grace_period;
+
+public class HeartbeatService
+{
+    private static final Logger logger = LoggerFactory.getLogger(HeartbeatService.class);
+
+    public static final HeartbeatService instance = new HeartbeatService();
+    private ScheduledExecutorPlus executor;
+    private long delay = 1;
+    private TimeUnit delayTimeUnit = TimeUnit.MINUTES;
+
+    private boolean started = false;
+    private Runnable heartbeat;
+
+    @VisibleForTesting
+    HeartbeatService()
+    {
+    }
+
+    @VisibleForTesting
+    void setHeartbeat(final Runnable heartbeat)
+    {
+        this.heartbeat = heartbeat;
+    }
+
+    @VisibleForTesting
+    void setDelay(long delay, TimeUnit delayTimeUnit)
+    {
+        if (delay > 0 && delayTimeUnit != null)
+        {
+            this.delay = delay;
+            this.delayTimeUnit = delayTimeUnit;
+        }
+    }
+
+    @VisibleForTesting
+    void setExecutor(final ScheduledExecutorPlus executor)
+    {
+        if (started)
+            throw new IllegalStateException("Can not set executor when service is started. Stop it first.");
+
+        this.executor = executor;
+    }
+
+    @VisibleForTesting
+    StartupChecksOptions getStartupChecksOptions()
+    {
+        return DatabaseDescriptor.getStartupChecksOptions();
+    }
+
+    public synchronized void start()
+    {
+        if (started)
+            return;
+
+        if (!getStartupChecksOptions().isEnabled(gc_grace_period))
+        {
+            logger.debug("Heartbeat service is disabled.");
+            return;
+        }
+
+        if (heartbeat == null)
+            heartbeat = new Heartbeat(getHeartbeatFile(getStartupChecksOptions().getConfig(gc_grace_period)));
+
+        if (executor == null)
+            executor = executorFactory().scheduled(false, "HeartbeatService");
+
+        if (executor.isShutdown())
+            throw new IllegalStateException("Executor to run heartbeats on is shut down!");
+
+        executor.scheduleWithFixedDelay(heartbeat, 0, delay, delayTimeUnit);
+
+        started = true;
+    }
+
+    public synchronized void stop() throws InterruptedException, TimeoutException
+    {
+        if (executor != null)
+            ExecutorUtils.shutdownNowAndWait(1L, TimeUnit.MINUTES, executor);
+
+        executor = null;
+        started = false;
+    }
+
+    private static class Heartbeat implements Runnable
+    {
+        private final File heartbeatFile;
+
+        public Heartbeat(File heartbeatFile)
+        {
+            this.heartbeatFile = heartbeatFile;
+        }
+
+        @Override
+        public void run()
+        {
+            FileUtils.write(heartbeatFile, Long.toString(Clock.Global.currentTimeMillis()));
```

Comment:
Should be timestamp written in the file rather human readable for easier manual verification / inspection? It is quite hard to grasp what was the last time the heartbeat file was written to when one see just a number only.